### ISSUE TITLE
Maize Model P Response Not Working

### ISSUE THREAD

**Issue by hut104** [APSIMInitiative/APSIMClassic#1736]
Tests of new Maize model with SoilP failed to find any response to P supply.  Some P calculations seem to be working in that P uptake seems to be taking place, but growth does not seem to be affected.

Tests have found that P response is still working in the old maize model and so the current work around is to use APSIM version 7.8 with the older maize model.  This also suggests that the current error may not be due to an error in SoilP.

**Comment by zur003** [comment]
It's been a year since this issue was opened, but the problem still remains. It's a nuisance for @EnliWang, who is working on changes in the SoilP model.

**Comment by EnliWang** [comment]
What were the main changes in the maize model from V7.9 to V7.10? My student used V7.9 and the model worked fine.

**Comment by zur003** [comment]
@EnliWang - @hol353 would know better than I, but if I recall correctly, the old Fortran maize model was completely replaced with a C++ version, which was previously known as MaizeZ.

**Comment by hol353** [comment]
Yep you are correct. The old Fortran maize model was replaced with a C++ model (the same code base as the Sorghum model I believe). The advice to our users who want a maize P response is to go back to the original maize model (version 7.9 I think).

**Comment by EnliWang** [comment]
We should fix this, otherwise the model loses its ability to simulate P response

**Comment by hol430** [comment]
Does this affect sorghum as well? The codebases are very similar and I've had to copy bugfixes from one to the other in the past.

**Comment by EnliWang** [comment]
Not sure, I don't sorghum simulations. I have simulated wheat, soybean and peanut. They are all responsive. Only maize is not.

**Comment by hut104** [comment]
Yes, there are multiple PhD students working in older versions of APSIM because of this problem.
There are also issues with SWIM and cultivars I think in the new maize model.

________________________________

**Comment by jbrider** [comment]
Is there any validation data for the P response in Maize?
I went looking for validation data sets for the older maize model and couldn't find any.

**Comment by EnliWang** [comment]
We have simulated wheat, maize, soybean and peanut in a rotation, will try to put these data in. In addition, a PhD student simulated maize in China, will put the data in some later time.

**Comment by hol430** [comment]
@hut104 do you remember what the problems with maize and swim are?

**Comment by hut104** [comment]
You may need to ask Sotirios.

________________________________

**Comment by jbrider** [comment]
will look at this soon.
@EnliWang is going to provide some simulations that are working in 7.9 that we can use to help test the results from 7.10.

**Comment by sarahcleary** [comment]
@Keith-Pembleton - will discuss with post doc around work conducted in this space.  Provided contact/feedback to @jbrider

**Comment by hol430** [comment]
After some investigation it appears that the maize model is P-aware, but the phosphorus variables aren't really used anywhere. I have hooked up the stress variables into some other stressful calculations and am now getting some P response. I'm not certain if this is all that is required or if some other variables need to be hooked up as well. I will make a pull request for this change and then see if more work is required.

For reference, here is the response I'm getting. These graphs come from a modified version of the "Maize P.apsim" example file shipped with apsim 7.9. I have 11 simulations, with varying amounts of P applied at sowing. I ran the file on apsim 7.9, 7.10 (before I applied my fix) and 7.10 (after applying my fix).

# APSIM 7.10 - Before Changes

![image](https://user-images.githubusercontent.com/36427516/81373371-f786df80-913f-11ea-88df-c5e5729af3b9.png)

# APSIM 7.10 - After Changes

![image](https://user-images.githubusercontent.com/36427516/81373378-fe155700-913f-11ea-8d66-4b6f31403f4d.png)

# APSIM 7.9

![image](https://user-images.githubusercontent.com/36427516/81373393-079ebf00-9140-11ea-8211-cada7f5fd37c.png)

**Comment by EnliWang** [comment]
This looks much better. If you let me know when this has been done, I will get the updated code and run my simulations and see what happens. Will feedback then.

**Comment by hol430** [comment]
Hi @EnliWang - this fix has been merged. If you're using an installed version, you can download an installer which includes this fix from the [downloads page](http://apsimdev.apsim.info/APSIM.Registration.Portal/Downloads.aspx). Otherwise, you can just update your source code.

**Comment by peter-devoil** [comment]
@EnliWang I have come across a units inconsistency with maize & sorghum's p_demand - they're sending out kg/ha, but soilP is expecting g/m2. So far the only symptom I can see is a large summary file full of warnings - I've yet to look further.

**Comment by EnliWang** [comment]
@peter-devoil Thanks Peter for the progress and for letting me know. I have not been able to get my old simulations working with APSIM 7.10. My maize plant does not seem to get enough P, i.e., simulated P uptake is only about half of measured P uptake.

@jbrider also, I tried to attach my simulation files, but too big. I will send via email.
